README.md
# Hanami::UJS

Unobtrusive JavaScript (UJS) for [Hanami](http://hanamirb.org).

## Status

[![Gem Version](http://img.shields.io/gem/v/hanami-ujs.svg)](https://badge.fury.io/rb/hanami-ujs)

## Installation

Add this line to your Hanami project's `Gemfile`:

```ruby
gem "hanami-ujs"
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install hanami-ujs

## Usage

### Setup

You have to add two lines to the application layout (eg. `apps/web/templates/application.html.erb`):

  1. `<%= csrf_meta_tags %>` inside `<head>`
  2. `<%= javascript "hanami-ujs" %>` the location is optional, but before `</body>` is a good spot.

### Events

Hanami UJS fires events to notify listeners that a certain operation happened:

  * `"ajax.before"`
  * `"ajax.complete"`

You can listen to these events with:

```js
(function() {
  var ajaxBeforeHandler = function(event) {
    console.log(event);
  };

  var ajaxCompleteHandler = function(event) {
    console.log(event);
  };

  document.addEventListener("ajax:before", ajaxBeforeHandler);
  document.addEventListener("ajax:complete", ajaxCompleteHandler);
})();
```

### AJAX Form

```erb
<%=
  form_for :search, "/search", remote: true do
    # ...

    submit "Search"
  end
%>
```

When the user will hit "Search" the form will be sent via AJAX.

## Development

After checking out the repo, run `bin/setup` to install dependencies. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/hanami/ujs.

## Acknowledgements

This gem JavaScript code is from the great [vanilla-ujs](https://rubygems.org/gems/vanilla-ujs) gem,
which is MIT licensed (Copyright (c) 2013 Łukasz Niemier). Thank you for your awesome work!

## Copyright

Copyright © 2018 Luca Guidi – Released under MIT License


CHANGELOG.md
# Hanami::UJS
Hanami Unobtrusive JavaScript (UJS)

## v0.1.0 - 2018-04-11

## v0.1.0.rc2 - 2018-04-06

## v0.1.0.rc1 - 2018-03-30

## v0.1.0.beta2 - 2018-03-23

## v0.1.0.beta1 - 2018-02-28
### Added
- [Luca Guidi] Added support for AJAX forms
- [Luca Guidi] Package as Hanami plugin
- [Marion Schleifer] Package JS code from `vanilla-ujs` gem


spec/integration/disable_spec.rb
# frozen_string_literal: true

RSpec.describe "Disable", type: :integration do
  it "successfully disable submit button" do
    with_project do
      generate "action web home#index --url=/"
      rewrite "apps/web/templates/home/index.html.erb", <<~CODE
        <%=
          form_for :search, "/search", "id": "my-form" do
            div do
              label "Term", for: :q
              text_field :q
            end

            div do
              submit "Search", "id": "submit-button", "data-disable-with": "Searching.."
            end
          end
        %>

        <div id="console"></div>
      CODE

      generate "action web search#index --url=/search --method=POST"

      rewrite "apps/web/templates/application.html.erb", <<~CODE
        <!DOCTYPE html>
        <html>
          <head>
            <title>Web</title>
            <%= favicon %>
            <%= csrf_meta_tags %>
          </head>
          <body>
            <%= yield %>
            <%= javascript "hanami-ujs", "home" %>
          </body>
        </html>
      CODE

      write "apps/web/assets/javascripts/home.js", <<~CODE
        (function() {
          function log(text) {
            append("console", "<p>" + text + "</p>");
          }

          function append(domID, html) {
            el = document.getElementById(domID);
            el.innerHTML += html;
          }

          var submitHandler = function(event) {
            event.preventDefault();

            window.setTimeout(function(){
              var submitButton = document.getElementById("submit-button");

              log("disabled: " + submitButton.disabled);
              log("label: " + submitButton.innerHTML);
            }, 1000);
          };

          var myForm = document.getElementById("my-form");
          myForm.addEventListener("submit", submitHandler);
        })();
      CODE

      server do
        visit "/"

        fill_in "Term", with: "Hanami"
        click_button "Search"

        expect(current_path).to eq("/")

        within "#console" do
          expect(page).to have_content("disabled: true")
          expect(page).to have_content("label: Searching..")
        end
      end
    end
  end

  private

  def with_project
    super("bookshelf", gems: { "hanami-ujs" => { path: Pathname.new(__dir__).join("..", "..").realpath.to_s } }) do
      replace_last "apps/web/application.rb", "# sessions :cookie, secret: ENV['WEB_SESSIONS_SECRET']", "sessions :cookie, secret: ENV['WEB_SESSIONS_SECRET']"
      bundle_install
      yield
    end
  end
end


spec/unit/hanami/ujs/version_spec.rb
# frozen_string_literal: true

RSpec.describe "Hanami::UJS::VERSION" do
  it "exposes version" do
    expect(Hanami::UJS::VERSION).to eq("0.1.0")
  end
end


spec/integration/submit_remote_form_spec.rb
# frozen_string_literal: true

RSpec.describe "Submit remote form", type: :integration do
  it "successfully sends AJAX request" do
    with_project do
      generate "action web home#index --url=/"
      rewrite "apps/web/templates/home/index.html.erb", <<~CODE
        <%=
          form_for :search, "/search", remote: true do
            div do
              label "Term", for: :q
              text_field :q
            end

            div do
              submit "Search"
            end
          end
        %>

        <div id="results">
        </div>

        <div id="console">
        </div>
      CODE

      generate "action web search#index --url=/search --method=POST"
      rewrite "apps/web/controllers/search/index.rb", <<~CODE
        module Web::Controllers::Search
          class Index
            include Web::Action
            expose :results

            def call(params)
              @results = [{ project: params.get(:search, :q) }]
              self.format = :json if xhr?(params.env)
            end

            private

            def xhr?(env)
              env["HTTP_X_REQUESTED_WITH"] == "XmlHttpRequest"
            end
          end
        end
      CODE

      write "apps/web/templates/search/index.json.erb", <<~CODE
        <%= results.to_json %>
      CODE

      rewrite "apps/web/templates/application.html.erb", <<~CODE
        <!DOCTYPE html>
        <html>
          <head>
            <title>Web</title>
            <%= favicon %>
            <%= csrf_meta_tags %>
          </head>
          <body>
            <%= yield %>
            <%= javascript "hanami-ujs", "home" %>
          </body>
        </html>
      CODE

      write "apps/web/assets/javascripts/home.js", <<~CODE
        (function() {
          function log(event) {
            append("console", "<p>" + event.type + "</p>");
            // console.log(event);
          }

          function append(domID, html) {
            el = document.getElementById(domID);
            el.innerHTML += html;
          }

          function reset(domID) {
            el = document.getElementById(domID);
            el.innerHTML = "";
          }

          var ajaxBeforeHandler = function(event) {
            log(event);
          };

          var ajaxCompleteHandler = function(event) {
            log(event);
            var data = JSON.parse(event.detail.responseText);

            reset("results");
            data.map( function(item) {
              append("results", "<p>" + item.project + "</p>");
            })
          };

          document.addEventListener("ajax:before", ajaxBeforeHandler);
          document.addEventListener("ajax:complete", ajaxCompleteHandler);
        })();
      CODE

      server do
        visit "/"

        fill_in "Term", with: "Hanami"
        click_button "Search"

        expect(current_path).to eq("/")

        within "#results" do
          expect(page).to have_content("Hanami")
        end

        within "#console" do
          expect(page).to have_content("ajax:before")
          expect(page).to have_content("ajax:complete")
        end
      end
    end
  end

  private

  def with_project
    super("bookshelf", gems: { "hanami-ujs" => { path: Pathname.new(__dir__).join("..", "..").realpath.to_s } }) do
      replace_last "apps/web/application.rb", "# sessions :cookie, secret: ENV['WEB_SESSIONS_SECRET']", "sessions :cookie, secret: ENV['WEB_SESSIONS_SECRET']"
      bundle_install
      yield
    end
  end
end


lib/hanami/ujs/ujs.js
document.addEventListener("click",function(t){var e,a;if(a=t.target,matches.call(a,"a[data-confirm], button[data-confirm], input[data-confirm]")){if(e=a.getAttribute("data-confirm"),!confirm(e))return t.stopPropagation(),t.stopImmediatePropagation(),t.preventDefault(),!1}else;},!1);var CSRF={token:function(){var t=document.querySelector('meta[name="csrf-token"]');return t&&t.getAttribute("content")},param:function(){var t=document.querySelector('meta[name="csrf-param"]');return t&&t.getAttribute("content")}},sameOrigin=function(t){var e,a=document.createElement("a");return a.href=t,e=a.href.split("/",3).join("/"),0===window.location.href.indexOf(e)};window.CSRF=CSRF,document.addEventListener("ajax:before",function(t){var e=CSRF.token(),a=t.detail;e&&a.setRequestHeader("X-CSRF-Token",e)}),document.addEventListener("submit",function(t){var e=CSRF.token(),a=CSRF.param(),n=t.target;if(matches.call(n,"form")){if(matches.call(n,"form[data-remote]"))return!0;if(!n.method||"GET"==n.method.toUpperCase())return!0;if(!sameOrigin(n.action))return!0;if(a&&e&&!n.querySelector("input[name="+a+"]")){var r=document.createElement("input");r.setAttribute("type","hidden"),r.setAttribute("name",a),r.setAttribute("value",e),n.appendChild(r)}return!0}}),document.addEventListener("click",function(t){var e,a;3!=t.which&&("submit"===(a=t.target).getAttribute("type")&&null!==a.form.querySelector(":invalid")||matches.call(a,"a[data-disable-with], button[data-disable-with], input[data-disable-with]")&&(e=a.getAttribute("data-disable-with"),a.value?a.value=e:a.innerHTML=e,setTimeout(function(){a.setAttribute("disabled","disabled")},0)))},!1);var VanillaUJS={formHasNoInputs:function(t){for(var e,a,n=0,r=t.elements,o=r.length;n<o;n++)if(e=r[n],a=e.nodeName.toUpperCase(),e.hasAttribute("name")&&!e.disabled&&("RADIO"!=a&&"CHECKBOX"!=a||e.checked))return!1;return!0}};document.addEventListener("submit",function(t){var e=t.target;if(matches.call(e,"form[data-remote]")){var a=e.action,n=(e.method||e.getAttribute("data-method")||"POST").toUpperCase(),r=new FormData(e),o=VanillaUJS.formHasNoInputs(e);if(CSRF.param()&&CSRF.token()?r[CSRF.param()]=CSRF.token():o&&(r=null),!LiteAjax.ajax({url:a,method:n,data:r,target:e}))return!0;t.preventDefault()}});var LiteAjax=function(){var LiteAjax={};return LiteAjax.options={method:"GET",url:window.location.href},LiteAjax.ajax=function(url,options){"object"==typeof url&&(options=url,url=void 0),(options=options||{}).accepts||(options.accepts="text/javascript, application/javascript, application/ecmascript, application/x-ecmascript"),url=url||options.url||location.href||"";var data=options.data,target=options.target||document,xhr=new XMLHttpRequest;xhr.addEventListener("load",function(){var responseType=xhr.getResponseHeader("content-type");"text/javascript; charset=utf-8"===responseType&&eval(xhr.response);var event=new CustomEvent("ajax:complete",{detail:xhr,bubbles:!0});target.dispatchEvent(event)}),"function"==typeof options.success&&xhr.addEventListener("load",function(t){xhr.status>=200&&xhr.status<300&&options.success(xhr)}),"function"==typeof options.error&&(xhr.addEventListener("load",function(t){(xhr.status<200||xhr.status>=300)&&options.error(xhr)}),xhr.addEventListener("error",function(t){options.error(xhr)})),xhr.open(options.method||"GET",url),xhr.setRequestHeader("X-Requested-With","XmlHttpRequest"),xhr.setRequestHeader("Accept","*/*;q=0.5, "+options.accepts),options.json&&(xhr.setRequestHeader("Content-type","application/json"),data=JSON.stringify(data));var beforeSend=new CustomEvent("ajax:before",{detail:xhr,bubbles:!0});return target.dispatchEvent(beforeSend),xhr.send(data),xhr},LiteAjax}();document.addEventListener("click",function(t){var e,a,n,r;if(0===t.button&&(e=t.target,matches.call(e,"a[data-method]"))){if(a=e.getAttribute("href"),n=e.getAttribute("data-method").toUpperCase(),r={},CSRF.param()&&CSRF.token()&&(r[CSRF.param()]=CSRF.token()),!(matches.call(e,"a[data-remote]")?function(t){return LiteAjax.ajax(t),!0}:function(t){var e,a,n;if("GET"==t.method)return!1;(e=document.createElement("form")).method="POST",e.action=t.url,e.style.display="none";for(n in t.data)Object.prototype.hasOwnProperty.call(t.data,n)&&((a=document.createElement("input")).setAttribute("type","hidden"),a.setAttribute("name",n),a.setAttribute("value",t.data[n]),e.appendChild(a));return"POST"!=t.method&&((a=document.createElement("input")).setAttribute("type","hidden"),a.setAttribute("name","_method"),a.setAttribute("value",t.method),e.appendChild(a)),document.body.appendChild(e),e.submit(),!0})({url:a,method:n,data:r,target:e}))return!0;t.preventDefault()}},!1);var matches=function(t){return t.matchesSelector||t.webkitMatchesSelector||t.mozMatchesSelector||t.oMatchesSelector||t.msMatchesSelector}(document.documentElement),CustomEvent=function(t,e){e=e||{bubbles:!1,cancelable:!1,detail:void 0};var a=document.createEvent("CustomEvent");return a.initCustomEvent(t,e.bubbles,e.cancelable,e.detail),a};CustomEvent.prototype=window.CustomEvent.prototype,window.CustomEvent=CustomEvent;

lib/hanami/ujs/assets/javascripts/hanami-ujs.js
//
// hanami-ujs (0.1.0)
// vanilla-ujs (1.3.0)
//
// Mon Feb 12 10:45:37 UTC 2018
//

//
// source: confirm.js
//
document.addEventListener('click', function (event) {
  var message, element;

  element = event.target;

  if (matches.call(element, 'a[data-confirm], button[data-confirm], input[data-confirm]')) {
    message = element.getAttribute('data-confirm');
    if (!confirm(message)) {
      event.stopPropagation();
      event.stopImmediatePropagation();
      event.preventDefault();
      return false;
    }

    return;
  }
}, false);

//
// source: csrf.js
//
var CSRF = {
  token: function () {
    var token = document.querySelector('meta[name="csrf-token"]');
    return token && token.getAttribute('content');
  },
  param: function () {
    var param = document.querySelector('meta[name="csrf-param"]');
    return param && param.getAttribute('content');
  }
};

var sameOrigin = function (url) {
  var a = document.createElement('a'), origin;
  a.href = url;
  origin = a.href.split('/', 3).join('/');

  return window.location.href.indexOf(origin) === 0;
};

window.CSRF = CSRF;

document.addEventListener('ajax:before', function (e) {
  var token = CSRF.token(), xhr = e.detail;
  if (token)
    xhr.setRequestHeader('X-CSRF-Token', token);
});

document.addEventListener('submit', function (e) {
  var token = CSRF.token(),
      param = CSRF.param(),
      form  = e.target;

  if (matches.call(form, 'form')) {
    if (matches.call(form, 'form[data-remote]'))
      return true;
    if (!form.method || form.method.toUpperCase() == 'GET')
      return true;
    if (!sameOrigin(form.action))
      return true;

    if (param && token && !form.querySelector('input[name='+param+']')) {
      var input = document.createElement('input');
      input.setAttribute('type', 'hidden');
      input.setAttribute('name', param);
      input.setAttribute('value', token);

      form.appendChild(input);
    }

    return true;
  }
});

//
// source: disable.js
//
document.addEventListener('click', function (event) {
  var message, element;

  // do not disable on right click. Work on left and middle click
  if (event.which == 3) {
    return;
  }

  element = event.target;

  // do not disable if the element is a submit button and its form has invalid input elements.
  // since failed validations prevent the form from being submitted, we would lock the form permanently
  // by disabling the submit button even though the form was never submitted

  if(element.getAttribute("type") === "submit" && element.form.querySelector(":invalid") !== null) {
    return;
  }

  if (matches.call(element, 'a[data-disable-with], button[data-disable-with], input[data-disable-with]')) {
    message = element.getAttribute('data-disable-with');
    if(!!element.value){
      element.value = message;
    }else{
      element.innerHTML = message;
    }
    // timeout is needed because Safari stops the submit if the button is immediately disabled
    setTimeout(function(){
      element.setAttribute('disabled', 'disabled');
    }, 0);
  }
}, false);

//
// source: form.js
//
document.addEventListener('submit', function(event) {

  var form = event.target;

  if (matches.call(form, 'form[data-remote]')) {
    var url = form.action;
    var method = (form.method || form.getAttribute('data-method') || 'POST').toUpperCase();
    var data = new FormData(form);

    if (CSRF.param() && CSRF.token()) {
      data[CSRF.param()] = CSRF.token();
    }

    if (LiteAjax.ajax({ url: url, method: method, data: data, target: form })){
      event.preventDefault();
    } else {
      return true;
    }
  }
});

//
// source: liteajax.js
//
var LiteAjax = (function () {
  var LiteAjax = {};

  LiteAjax.options = {
    method: 'GET',
    url: window.location.href
  };

  LiteAjax.ajax = function (url, options) {
    if (typeof url == 'object') {
      options = url;
      url = undefined;
    }

    options = options || {};

    if(!options.accepts) {
      options.accepts = 'text/javascript, application/javascript, ' +
                        'application/ecmascript, application/x-ecmascript';
    }

    url = url || options.url || location.href || '';
    var data = options.data;
    var target = options.target || document;
    var xhr = new XMLHttpRequest();

    xhr.addEventListener('load', function () {
      var responseType = xhr.getResponseHeader('content-type');
      if(responseType === 'text/javascript; charset=utf-8') {
        eval(xhr.response);
      }

      var event = new CustomEvent('ajax:complete', {detail: xhr, bubbles: true});
      target.dispatchEvent(event);
    });

    if (typeof options.success == 'function')
      xhr.addEventListener('load', function (event) {
        if (xhr.status >= 200 && xhr.status < 300)
          options.success(xhr);
      });

    if (typeof options.error == 'function') {
      xhr.addEventListener('load', function (event) {
        if (xhr.status < 200 || xhr.status >= 300)
          options.error(xhr);
      });
      xhr.addEventListener('error', function (event) {
        options.error(xhr);
      });
    }

    xhr.open(options.method || 'GET', url);
    xhr.setRequestHeader('X-Requested-With', 'XmlHttpRequest');
    xhr.setRequestHeader('Accept', '*/*;q=0.5, ' + options.accepts);

    if(options.json) {
      xhr.setRequestHeader('Content-type', 'application/json');
      data = JSON.stringify(data);
    }

    var beforeSend = new CustomEvent('ajax:before', {detail: xhr, bubbles: true});
    target.dispatchEvent(beforeSend);
    xhr.send(data);

    return xhr;
  };

  return LiteAjax;
})();

//
// source: method.js
//
document.addEventListener('click', function(event) {
  var element, url, method, data, handler;

  // Only left click allowed. Firefox triggers click event on right click/contextmenu.
  if (event.button !== 0) {
    return;
  }

  element = event.target;

  if (matches.call(element, 'a[data-method]')) {
    url = element.getAttribute('href');
    method = element.getAttribute('data-method').toUpperCase();
    data = {};

    if (CSRF.param() && CSRF.token()) {
      data[CSRF.param()] = CSRF.token();
    }

    if (matches.call(element, 'a[data-remote]')) {
      handler = xhr;
    } else {
      handler = submit;
    }

    if (handler({ url: url, method: method, data: data, target: element })) {
      event.preventDefault();
    } else {
      return true;
    }
  }

  function submit(options) {
    var form, input, param;

    if (options.method == 'GET') {
      return false;
    }

    form = document.createElement('form');
    form.method = 'POST';
    form.action = options.url;
    form.style.display = 'none';

    for (param in options.data) {
      if (Object.prototype.hasOwnProperty.call(options.data, param)) {
        input = document.createElement('input');
        input.setAttribute('type', 'hidden');
        input.setAttribute('name', param);
        input.setAttribute('value', options.data[param]);
        form.appendChild(input);
      }
    }

    if (options.method != 'POST') {
      input = document.createElement('input');
      input.setAttribute('type', 'hidden');
      input.setAttribute('name', '_method');
      input.setAttribute('value', options.method);
      form.appendChild(input);
    }

    document.body.appendChild(form);
    form.submit();
    return true;
  }

  function xhr(options) {
    LiteAjax.ajax(options);
    return true;
  }
}, false);

//
// source: polyfills.js
//
var matches = (function(doc) {
  return doc.matchesSelector ||
    doc.webkitMatchesSelector ||
    doc.mozMatchesSelector ||
    doc.oMatchesSelector ||
    doc.msMatchesSelector;
})(document.documentElement);

var CustomEvent = function (event, params) {
  params = params || {bubbles: false, cancelable: false, detail: undefined};
  var evt = document.createEvent('CustomEvent');
  evt.initCustomEvent(event, params.bubbles, params.cancelable, params.detail);
  return evt;
};

CustomEvent.prototype = window.CustomEvent.prototype;

window.CustomEvent = CustomEvent;
